Improve `singleton-comparison` suggestion message

ethan-leba opened the pull request:
<!--

Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to pylint in this document:
https://github.com/PyCQA/pylint/blob/master/doc/development_guide/contribute.rst#repository
-->

## Description
Singleton-comparison will now suggest `is` for boolean constants, as well as a message for truthiness/falsiness in case the user intended this, whereas the message would only suggest checking truthiness/falsiness before. The message will now also display the actual code in question, instead of an `expr` stand-in.

## Type of Changes
<!-- Leave the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|      | :sparkles: New feature |
|      | :hammer: Refactoring  |
|      | :scroll: Docs |

Pierre-Sassoulas commented on pylint/checkers/base.py:
```diff
@@ -2345,7 +2345,9 @@ class ComparisonChecker(_BasicChecker):
             if not negative_check:
-                suggestion = "just 'expr'"
+                suggestion = "just '{}'".format(
+                    "expr" if utils.is_test_condition(root_node) else "bool(expr)"
```
I know it's almost a different feature than what you're doing here, but would it be possible to do bool({expr}) by passing expr to `_check_singleton_comparison` ?  Also most of the time it should be just "expr" and the cast to bool is implicit ? I think it would make for a lot better message. what do you think ?

ethan-leba replied:
From looking at your comment below, by `bool({expr})` do you mean show the actual expression in question instead of `"expr"`?

ethan-leba replied:
> Also most of the time it should be just "expr" and the cast to bool is implicit ?

What do you mean by this?

Pierre-Sassoulas replied:
Yes sorry I meant the actual expression. For example if the real expression is "x", I think we can also output `Consider replacing 'x == True' to just 'x' instead` and not `Consider replacing 'x == True' to just 'bool(x)'` when python would call cast to a boolean ?

ethan-leba replied:
That's what this PR was originally addressing -- situations where Python wouldn't cast to boolean, i.e. assigning to a variable as mentioned in the description. Am I understanding you correctly? But yes, I agree that the above would make for a much better error message, I'll add that in 👍

Pierre-Sassoulas commented on tests/functional/s/singleton_comparison.txt:
```diff
@@ -1,8 +1,12 @@
 singleton-comparison:15::Comparison to None should be 'expr is not None'
+singleton-comparison:16::Comparison to True should be just 'expr'
+singleton-comparison:17::Comparison to False should be 'expr'
+singleton-comparison:18::Comparison to True should be just 'expr'
```
To make my last suggestion clearer, it could be:
```suggestion
singleton-comparison:18::Consider replacing 'x == True' to just 'x'
```

Pierre-Sassoulas reviewed (approved):
I like this change. I don't like that the "utils.py" file is getting bigger because for me "utils" are where we we put things we can't name properly. But I don't have a smart name to suggest, so... Can it be used elsewhere ?

ethan-leba commented:
Thanks for the quick review! In regards to the utils file, I'm not too sure what the best course of action there is either. I don't think it makes sense to keep those functions in the len-checker file if we're importing it in a bunch of different places though. Maybe the utils file could be split into a package and categorized in another PR? Possibly as a stopgap we could add a utils module to the new refactoring package for now

Pierre-Sassoulas commented:
> Maybe the utils file could be split into a package and categorized in another PR?
Yeah, probably. We've started doing that before. Maybe those utility function should be in Astroid instead. Anyway I have no good alternative ideas and this is not the main focus of the PR anyway. Moving them to a more generic file like you did is good enough for now :)

Pierre-Sassoulas commented on tests/functional/s/singleton_comparison.txt:
```diff
@@ -1,8 +1,12 @@
 singleton-comparison:4::Comparison to None should be 'expr is None'
-singleton-comparison:5::Comparison to True should be just 'expr'
+singleton-comparison:5::Comparison to True should be just 'bool(expr)'
```
I think `x = x==True` and bool(x) are not equivalent, for example if x = ["1"]:

```
>>> x = ["1"]
>>> y = x==True
False
>>> y = bool(x)
True
```

ethan-leba replied:
Hmm, you're right about that. The current suggestion is incorrect as well though:
```
>>> x = ["1"]
>>> if x == True:
...     print("hello")
...
>>> if x: (suggested)
...     print("hello")
...
hello
>>> x = ["1"] == True
>>> x
False
>>> x = ["1"] (suggested)
>>> x
['1']
```
I think the suggestion should be `expr is True/False` instead of `expr/not expr`?

Pierre-Sassoulas replied:
Yeah, I think the problem if there is a comparison between a list and True is that the code is so wrong that we can't know what the fix is supposed to be. Maybe the spirit of the warning is that x==True is wrong whatever the value of x might be. I have a hard time thinking of a good message though...  "Don't mix boolean and other types" ? "Consider replacing comparison to 'True' by an appropriate call to bool()" ? "Suspicious comparison to True, consider using bool(x) or a typed boolean" ?

ethan-leba replied:
How about `Use '{expr} is True' to check for the singleton value True, or '{expr}' to test for truthiness`?

Pierre-Sassoulas replied:
Very nice and imho understandable by everyone. Let's use that one :) !

ethan-leba commented:
Not in the scope of this current PR, but should we be checking for Enum members as well for `singleton-comparison`?

ethan-leba commented:
This PR should be ready for re-review now  @Pierre-Sassoulas 🙂

Pierre-Sassoulas commented on tests/functional/s/singleton_comparison.txt:
```diff
@@ -1,8 +1,12 @@
-singleton-comparison:13::Comparison to True should be just 'not expr'
-singleton-comparison:14::Comparison to False should be 'expr'
-singleton-comparison:15::Comparison to None should be 'expr is not None'
+singleton-comparison:4::Comparison 'x == None' should be 'x is None'
```
👌

Pierre-Sassoulas reviewed (approved):
I have some nitpicky comments but the result is awesome, thank you !

ethan-leba commented:
Added in the changes, let me know if there's anything else that needs work before merging! Also, for my [previous PR](https://github.com/PyCQA/pylint/pull/3644) I think I might've placed documentation incorrectly, since I added to the 'What's New' and 'ChangeLog' for `v2.6` after it had already been merged. Should I move this documentation to the `v2.7` section?

Pierre-Sassoulas commented:
Thank you for this merge request, this message is most useful for python's beginners; so a very clear message is helping a lot.
